test/unit/Types/ScalarTypesResolverTest.php

```diff
@@ -0,0 +1,210 @@
+                'expectedValue' => DateTime::createFromFormat('Y-m-d', '2020-07-31'),
+            ],
+            'deserialize_date_time' => [
+                'deserialize' => true,
+                'id' => 2,
+                'value' => '2020-07-31 15:35:11',
+                'expectedValue' => new DateTime('2020-07-31 15:35:11'),
+            ],
+            'deserialize_byte' => [
+                'deserialize' => true,
+                'id' => 3,
+                'value' => 'VGhpcyBpcyBhbiBlbmNvZGVkIHN0cmluZw==',
+                'expectedValue' => 'This is an encoded string',
+            ],
+            'deserialize_php_type_bool' => [
+                'deserialize' => true,
+                'id' => 11,
+                'value' => '',
+                'expectedValue' => false,
+            ],
+            'deserialize_php_type_float' => [
+                'deserialize' => true,
+                'id' => 5,
+                'value' => '1.1qwerty',
+                'expectedValue' => 1.1,
+            ],
+            'deserialize_php_type_int' => [
+                'deserialize' => true,
+                'id' => 10,
+                'value' => 'q1werty',
+                'expectedValue' => 0,
+            ],
+        ];
+    }
+
+    /**
+     * @param mixed $value
+     * @param mixed $expectedValue
+     *
+     * @dataProvider convertDeserializeDataProvider
+     */
+    public function testConvertDeserializesValue(bool $deserialize, int $id, $value, $expectedValue): void
+    {
+        $convertedValue = $this->scalarTypeResolver->convert($deserialize, $id, $value);
+        Assert::assertEquals($expectedValue, $convertedValue);
+    }
+
+    /**
+     * @return array|mixed[]
+     *
+     * @throws DatetimeException
+     */
+    public function convertSerializeDataProvider(): array
+    {
+        return [
+            'serialize_null' => [
+                'deserialize' => false,
+                'id' => 1,
+                'value' => null,
+                'expectedValue' => null,
+            ],
+            'serialize_date' => [
+                'deserialize' => false,
+                'id' => 1,
+                'value' => DateTime::createFromFormat('Y-m-d', '2020-07-31'),
+                'expectedValue' => '2020-07-31',
+            ],
+            'serialize_date_time' => [
+                'deserialize' => false,
+                'id' => 2,
+                'value' => new DateTime('2020-07-31 15:35:11'),
+                'expectedValue' => '2020-07-31T15:35:11+00:00',
+            ],
+            'serialize_byte' => [
+                'deserialize' => false,
+                'id' => 3,
+                'value' => 'This is an encoded string',
+                'expectedValue' => 'VGhpcyBpcyBhbiBlbmNvZGVkIHN0cmluZw==',
+            ],
+            'serialize_value_without_serializer_function' => [
+                'deserialize' => false,
+                'id' => 11,
+                'value' => 'q1werty',
```

Comment:
Same here, let's check first for an expected value, a string containing only an integer.
If someone makes API's declaring fields of number type and sends 'q1werty' there, let them suffer the consequences, no guarantees from us.